gt/main.c

```diff
@@ -890,14 +795,349 @@ print_unsent_policy(struct ggu_policy *policy)
+		if (ggu_pkt->buf == NULL)
+			continue;
+
+		if (pkt_info->outer_ethertype != ggu_pkt->ipaddr.proto)
+			continue;
+
+		if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+			struct ipv4_hdr *ipv4_hdr =
+				(struct ipv4_hdr *)pkt_info->outer_l3_hdr;
+			if (ggu_pkt->ipaddr.ip.v4.s_addr ==
+					ipv4_hdr->src_addr)
+				return ggu_pkt;
+		} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+			struct ipv6_hdr *ipv6_hdr =
+				(struct ipv6_hdr *)pkt_info->outer_l3_hdr;
+			if (ipv6_addrs_equal(ipv6_hdr->src_addr,
+					ggu_pkt->ipaddr.ip.v6.s6_addr))
+				return ggu_pkt;
+		}
+	}
+	return NULL;
+}
+
+static void
+prep_notify_pkt(struct ggu_notify_pkt *ggu_pkt)
+{
+	/*
+	 * Complete the packet fields that can only be done
+	 * when the packet is ready to be transmitted.
+	 */
+	struct udp_hdr *notify_udp;
+
+	if (ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv4) {
+		struct ipv4_hdr *notify_ipv4 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv4_hdr *,
+				ggu_pkt->buf->l2_len);
+		notify_ipv4->total_length = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len);
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv4[1];
+		notify_udp->dgram_cksum =
+			rte_ipv4_phdr_cksum(notify_ipv4,
+				ggu_pkt->buf->ol_flags);
+	} else if (likely(ggu_pkt->ipaddr.proto == ETHER_TYPE_IPv6)) {
+		struct ipv6_hdr *notify_ipv6 =
+			rte_pktmbuf_mtod_offset(ggu_pkt->buf,
+				struct ipv6_hdr *, ggu_pkt->buf->l2_len);
+		notify_ipv6->payload_len = rte_cpu_to_be_16(
+			ggu_pkt->buf->data_len - ggu_pkt->buf->l2_len -
+			sizeof(struct ipv6_hdr));
+
+		/* Offload the UDP checksum. */
+		notify_udp = (struct udp_hdr *)&notify_ipv6[1];
+		notify_udp->dgram_cksum =
+			rte_ipv6_phdr_cksum(notify_ipv6,
+				ggu_pkt->buf->ol_flags);
+	} else {
+		rte_panic("Unexpected condition: gt at lcore %u sending notification packet to Gatekeeper server with unknown IP version %hu\n",
+			rte_lcore_id(), ggu_pkt->ipaddr.proto);
+	}
+
+	notify_udp->dgram_len =
+		rte_cpu_to_be_16((uint16_t)(ggu_pkt->buf->data_len -
+			ggu_pkt->buf->l2_len - ggu_pkt->buf->l3_len));
+}
+
+static void
+send_notify_pkt(struct gt_config *gt_conf, struct gt_instance *instance,
+	struct ggu_notify_pkt *ggu_pkt)
+{
+	prep_notify_pkt(ggu_pkt);
+
+	if (rte_eth_tx_burst(gt_conf->net->front.id,
+			instance->tx_queue, &ggu_pkt->buf, 1) != 1) {
+		print_unsent_policies(ggu_pkt);
+		rte_pktmbuf_free(ggu_pkt->buf);
+	}
+
+	ggu_pkt->buf = NULL;
+	instance->num_ggu_pkts--;
+}
+
+/* Send all saved policy decision notification packets being buffered. */
+static void
+flush_notify_pkts(struct gt_config *gt_conf, struct gt_instance *instance)
+{
+	unsigned int max_pkts = gt_conf->max_ggu_notify_pkts;
+	struct rte_mbuf *bufs[max_pkts];
+	int num_to_send = 0;
+	int num_sent;
+	int sent_count = 0;
+	unsigned int i;
+
+	if (instance->ggu_pkts == NULL || instance->num_ggu_pkts == 0)
+		return;
+
+	for (i = 0; i < max_pkts; i++) {
+		if (instance->ggu_pkts[i].buf == NULL)
```

Comment:
Style: 

1. Define variable `struct ggu_notify_pkt *ggu_pkt;`;
2. Set the variable at the beginning of the loop: `ggu_pkt = &instance->ggu_pkts[i];`;
3. Simplify the expressions in the loop.

Steps 2 and 3 apply to the next loop as well.